src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -45,11 +48,16 @@ public void OnBeginRequest(HttpContext httpContext, long timestamp)
                 this.client.Initialize(requestTelemetry);
                 requestTelemetry.Start(timestamp);
                 httpContext.Features.Set(requestTelemetry);
-
                 IHeaderDictionary responseHeaders = httpContext.Response?.Headers;
-                if (responseHeaders != null && !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey) && !responseHeaders.ContainsKey(RequestResponseHeaders.TargetInstrumentationKeyHeader))
+                if (responseHeaders != null &&
+                    !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey) &&
+                    (!responseHeaders.ContainsKey(RequestResponseHeaders.RequestContextHeader) || HttpHeadersUtilities.ContainsRequestContextKeyValue(responseHeaders, RequestResponseHeaders.RequestContextTargetKey)))
```

Comment:
So the helper method does an iteration through all of the values, did the old way do a hash lookup?  Which is less expensive?